### PR TITLE
identity: Introduce ConflictResolver interface

### PR DESCRIPTION
**Pull request by mpalmi**
### Description

This PR introduces a new interface for conflict resolution of duplicate Identity artifacts. The initial implementation just reorganizes the code to use the interface with no behavior change.

The interface is intended to provide a minimal touchpoint for implementing new conflict resolution behavior. Since those changes will also introduce significant testcases, the aim here is to merge the new interface and ensure the current code works as intended (according to existing tests).

### TODO only if you're a HashiCorp employee
- [x] **Backport Labels:** If this fix needs to be backported, use the appropriate `backport/` label that matches the desired release branch. Note that in the CE repo, the latest release branch will look like `backport/x.x.x`, but older release branches will be `backport/ent/x.x.x+ent`.
    - [x] **LTS**: If this fixes a critical security vulnerability or [severity 1](https://www.hashicorp.com/customer-success/enterprise-support) bug, it will also need to be backported to the current [LTS versions](https://developer.hashicorp.com/vault/docs/enterprise/lts#why-is-there-a-risk-to-updating-to-a-non-lts-vault-enterprise-version) of Vault. To ensure this, use **all** available enterprise labels.
- [x] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [x] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [x] **RFC:** If this change has an associated RFC, please link it in the description.
- [x] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

### CODE REVIEW

File: vault/identity_store_util.go [hunk 1]
```diff
@@ -137,15 +143,12 @@ func (i *IdentityStore) loadGroups(ctx context.Context) error {
 			nsCtx := namespace.ContextWithNamespace(ctx, ns)
 
 			// Ensure that there are no groups with duplicate names
-			groupByName, err := i.MemDBGroupByName(nsCtx, group.Name, false)
+			groupByName, err := i.MemDBGroupByName(nsCtx, group.Name, true)
```
Comment (mpalmi) [inline vault/identity_store_util.go:146]: I think the clone here is correct even without renames, right? Do we want to do this in a separate PR with some docs about why?

File: vault/identity_store_util.go [hunk 2]
```diff
@@ -638,9 +646,8 @@ func (i *IdentityStore) upsertEntityInTxn(ctx context.Context, txn *memdb.Txn, e
 		}
 
 		if strutil.StrListContains(aliasFactors, i.sanitizeName(alias.Name)+alias.MountAccessor) {
-			i.logger.Warn(errDuplicateIdentityName.Error(), "alias_name", alias.Name, "mount_accessor", alias.MountAccessor, "local", alias.Local, "entity_name", entity.Name, "action", "delete one of the duplicate aliases")
```
Comment (mpalmi) [inline vault/identity_store_util.go:641]: Hrm, this doesn't quite match up with the ResolveAliases log line...

File: vault/identity_store_conflicts.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// pre-existing Identity artifact is found with the same factors as a new one.
+type errorResolver struct {
+	logger log.Logger
+}
+
+// ResolveEntities logs a warning message when a pre-existing Entity is found
+// and returns a duplicate name error, which should be handled by the caller by
+// putting the system in case-sensitive mode.
+func (r *errorResolver) ResolveEntities(ctx context.Context, existing, duplicate *identity.Entity) error {
+	if existing == nil {
```
Comment (banks) [inline vault/identity_store_conflicts.go:31]: Should we also `nil` check duplicate before we access `duplicate.Name` below?

**Review by banks — DISMISSED** [review]
LGTM Mike.

File: vault/identity_store_conflicts.go (hunk 3 above)
Comment (mpalmi) [reply vault/identity_store_conflicts.go:36]: This is actually kind of  nice because it allows us to search for different-case matches in case-sensitive mode I think. We might want to consider if this is a smart design choice, but I did add a comment in the interface about search flexibility.

File: vault/identity_store_conflicts.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		return nil
+	}
+
+	r.logger.Warn(errDuplicateIdentityName.Error(),
+		"entity_name", duplicate.Name,
+		"duplicate_of_id", existing.ID,
```
[the remark this answers is not in the document]
Comment (mpalmi) [reply vault/identity_store_conflicts.go:43]: I updated this for the time being since it was a quick update.

File: vault/identity_store_conflicts.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+import (
+	"context"
+	"errors"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/helper/identity"
+)
+
+var errDuplicateIdentityName = errors.New("duplicate identity name")
```
Comment (davidadeleon) [inline vault/identity_store_conflicts.go:14]: Looks like this is already declared in `identity_store_util`. I think it makes more sense for this to live here if this is where we're going to be making the determination that we have duplicates though so we can probably just remove it from there.

Comment (mpalmi) [reply vault/identity_store_conflicts.go:14]: Thanks @davidadeleon! Looks like vsCode didn't save my changes >.<

**Review by banks — COMMENTED** [review]
Super minor thread on the clone thing but rest is looking good.

File: vault/identity_store_conflicts.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	r.logger.Warn(errDuplicateIdentityName.Error(),
+		"entity_name", duplicate.Name,
+		"duplicate_of_name", existing.Name,
+		"duplicate_of_id", existing.ID,
+		"action", "merge the duplicate entities into one")
+
+	return errDuplicateIdentityName
+}
+
+// Resolveidentity.Groups logs a warning message when a pre-existing identity.Group is found and
```
Comment (mpalmi) [inline vault/identity_store_conflicts.go:49]:
```suggestion
// ResolveGroups logs a warning message when a pre-existing identity.Group is found and
```

**Review by banks — APPROVED** [review]
Great! LGTM
